[robust_mutex.cpp:476] | (Runtime) Unit Tests | runtime-unit-tests / runtime_core [bh_p150b_civ2]

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [CI] MetalContextTest.CreateSiliconInstance deadlocks on CHIP_IN_USE_0_PCIe mutex, causing job timeout on bh_p150b_civ2
<!-- AUTO-TRIAGE-SUMMARY-END -->

**Workflow:** (Runtime) Unit Tests
**Job:** runtime-unit-tests / runtime_core [bh_p150b_civ2]

### Regression timeline

- **Last passing:** commit `1769bd090998` on 2026-07-12T05:08:47Z — https://github.com/tenstorrent/tt-metal/actions/runs/29180740435
- **First failing:** commit `52c1d0e55474` on 2026-07-15T05:04:22Z — https://github.com/tenstorrent/tt-metal/actions/runs/29390511828/job/87483135283
- **Most recent failing:** commit `3218270556c5` on 2026-07-20T22:16:30Z — https://github.com/tenstorrent/tt-metal/actions/runs/29755743018/job/88486927475

### Failing job URLs (last 2 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/29755743018/job/88486927475
- https://github.com/tenstorrent/tt-metal/actions/runs/29718833976/job/88306091359

### Failing test path(s)

- `./build/test/tt_metal/unit_tests_context --gtest_filter='-MetalContextIntegrationTest.Legacy'`
- Hanging test: `MetalContextTest.CreateSiliconInstance`

### Error excerpt (terminal failure from logs)

**Run 1 (2026-07-16):**
```
2026-07-16 07:55:53.657 | warning  |             UMD | Waiting for lock 'CHIP_IN_USE_0_PCIe' which is currently held by thread TID: 604, PID: 604 (robust_mutex.cpp:476)
##[error]The action 'runtime_core [bh_p150b_civ2]' has timed out after 12 minutes.
```

**Run 2 (2026-07-15):**
```
2026-07-15 22:39:02.555 | warning  |             UMD | Waiting for lock 'CHIP_IN_USE_0_PCIe' which is currently held by thread TID: 621, PID: 621 (robust_mutex.cpp:476)
##[error]The action 'runtime_core [bh_p150b_civ2]' has timed out after 12 minutes.
```

### Reproduction steps

1. On a Blackhole (bh_p150b_civ2) device, run the unit context tests:
   ```bash
   export TT_METAL_HOME=/work
   export ARCH_NAME=blackhole
   export PYTHONPATH=/work
   export LD_LIBRARY_PATH=/work/build/lib
   ./build/test/tt_metal/unit_tests_context --gtest_filter='-MetalContextIntegrationTest.Legacy'
   ```
2. Wait for the `MetalContextTest` suite to begin after `MetalEnv` tests complete.
3. Observe that `MetalContextTest.CreateSiliconInstance` hangs — it attempts to acquire the `CHIP_IN_USE_0_PCIe` robust mutex, which is already held by the same process (self-deadlock or unreleased lock from the preceding `MetalEnv` test suite).

### Notes

- The test suite `MetalEnv` (18 tests) completes successfully, but leaves the `CHIP_IN_USE_0_PCIe` PCIe mutex locked. The immediately following test `MetalContextTest.CreateSiliconInstance` then tries to open the device and deadlocks waiting on that same mutex held by its own process/thread.
- The lock holder PID equals the waiting PID in both runs (PID 604 in run 1, PID 621 in run 2), indicating a reentrant/unreleased lock rather than cross-process contention.
- Both runs also emit two identical `MetalEnv has 1 outstanding reference(s) at teardown or fork` errors before the deadlock, suggesting a resource cleanup issue at the end of the `MetalEnv` test suite that leaves the device lock in an acquired state.
- **Determinism check:**
  - Run 1 terminal error: `Waiting for lock 'CHIP_IN_USE_0_PCIe' which is currently held by thread TID: 604, PID: 604` → 12-minute job timeout
  - Run 2 terminal error: `Waiting for lock 'CHIP_IN_USE_0_PCIe' which is currently held by thread TID: 621, PID: 621` → 12-minute job timeout
  - Same lock name, same self-deadlock pattern, same timeout in both runs.

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: (Runtime) Unit Tests`
`Auto-triage-job-name: runtime-unit-tests / runtime_core [bh_p150b_civ2]`
<!-- AUTO-TRIAGE-METADATA-END -->

<!-- AUTO-TRIAGE-INFO-START -->
## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info
<!-- AUTO-TRIAGE-INFO-END -->

<!-- PIPELINE-REPORT-START -->
## Pipeline report
Analysis report: https://github.com/tenstorrent/tt-metal/actions/runs/29460595023
Slack notification: https://tenstorrent.slack.com/archives/C0B1F8Z7GQM/p1784161131100439?thread_ts=1784160487.320479&cid=C0B1F8Z7GQM
<!-- PIPELINE-REPORT-END -->

--- a/tt_metal/impl/context/metal_env.cpp
+++ b/tt_metal/impl/context/metal_env.cpp
@@ -94,6 +94,10 @@ MetalEnvImpl::~MetalEnvImpl() {
         std::lock_guard<std::mutex> lock(s_registry_mutex_);
         s_registry_.erase(this);
     }
+    // NOTE: the env-owned context (see ensure_context_registered) is intentionally torn down earlier, in
+    // MetalEnv::~MetalEnv(), before the MetalEnv::impl_ unique_ptr is reset. MetalContext::teardown()
+    // routes get_cluster()/rtoptions() back through *env_ (the outer MetalEnv), whose impl_ is already
+    // nullptr by the time ~MetalEnvImpl runs, so destroying it here would dereference that null impl_.
     check_use_count_zero();
     teardown_fabric_objects();
     cluster_.reset();
@@ -376,6 +380,20 @@ bool MetalEnvImpl::consume_force_reinit() {
 
 // ─── Control plane ────────────────────────────────────────────────────────────
 
+int MetalEnvImpl::ensure_context_registered(MetalEnv& env) {
+    if (!registered_context_id_.has_value()) {
+        registered_context_id_ = MetalContext::create_instance(env).get();
+    }
+    return *registered_context_id_;
+}
+
+void MetalEnvImpl::teardown_registered_context() {
+    if (registered_context_id_.has_value()) {
+        MetalContext::destroy_instance(/*check_device_count=*/false, ContextId{*registered_context_id_});
+        registered_context_id_.reset();
+    }
+}
+
 tt::tt_fabric::ControlPlane& MetalEnvImpl::get_control_plane() {
     std::lock_guard<std::mutex> lock(control_plane_mutex_);
     if (!control_plane_) {
@@ -591,7 +609,13 @@ void MetalEnvImpl::teardown_fabric_objects() {
 
 MetalEnv::MetalEnv(MetalEnvDescriptor descriptor) : impl_(std::make_unique<MetalEnvImpl>(std::move(descriptor))) {}
 
-MetalEnv::~MetalEnv() { this->impl_.reset(); }
+MetalEnv::~MetalEnv() {
+    // Destroy the env-owned MetalContext (if any) while impl_ is still valid: MetalContext::teardown()
+    // reaches back through *env_ -> MetalEnv::impl_, which unique_ptr::reset() nulls before running
+    // ~MetalEnvImpl. See ensure_context_registered / teardown_registered_context.
+    impl_->teardown_registered_context();
+    this->impl_.reset();
+}
 
 const MetalEnvDescriptor& MetalEnv::get_descriptor() const { return impl_->get_descriptor(); }
 
@@ -614,8 +638,14 @@ float MetalEnv::get_eps() const { return impl_->get_hal().get_eps(); }
 float MetalEnv::get_nan() const { return impl_->get_hal().get_nan(); }
 float MetalEnv::get_inf() const { return impl_->get_hal().get_inf(); }
 
-tt::tt_fabric::ControlPlane& MetalEnv::get_control_plane() { return impl_->get_control_plane(); }
-distributed::SystemMesh& MetalEnv::get_system_mesh() { return impl_->get_system_mesh(); }
+tt::tt_fabric::ControlPlane& MetalEnv::get_control_plane() {
+    impl_->ensure_context_registered(*this);
+    return impl_->get_control_plane();
+}
+distributed::SystemMesh& MetalEnv::get_system_mesh() {
+    impl_->ensure_context_registered(*this);
+    return impl_->get_system_mesh();
+}
 std::shared_ptr<distributed::MeshDevice> MetalEnv::create_mesh_device(
     const distributed::MeshDeviceConfig& config,
     size_t l1_small_size,
@@ -626,7 +656,12 @@ std::shared_ptr<distributed::MeshDevice> MetalEnv::create_mesh_device(
     size_t worker_l1_size) {
     // Associate a context ID for the mesh device's dependencies to easily access the MetalContext::instance(contextId)
     // TODO: Remove this and directly pass in the MetalEnv reference
-    ContextId context_id = MetalContext::create_instance(*this);
+    // If the control plane / system mesh was already accessed, the env owns a registered context; reuse it
+    // (its lifetime is tied to the env). Otherwise create one here and let the mesh device tear it down on
+    // close, preserving the legacy create_mesh_device-only behavior.
+    const bool env_owns_context = impl_->has_registered_context();
+    ContextId context_id =
+        env_owns_context ? ContextId{impl_->ensure_context_registered(*this)} : MetalContext::create_instance(*this);
     auto mesh_device = distributed::MeshDeviceImpl::create(
         context_id,
         config,
@@ -636,7 +671,9 @@ std::shared_ptr<distributed::MeshDevice> MetalEnv::create_mesh_device(
         dispatch_core_config,
         l1_bank_remap,
         worker_l1_size);
-    mesh_device->impl().set_destroy_metal_context_instance_on_close(true);
+    if (!env_owns_context) {
+        mesh_device->impl().set_destroy_metal_context_instance_on_close(true);
+    }
     return mesh_device;
 }
 
@@ -648,7 +685,9 @@ std::shared_ptr<distributed::MeshDevice> MetalEnv::create_unit_mesh_device(
     const DispatchCoreConfig& dispatch_core_config,
     ttsl::Span<const std::uint32_t> l1_bank_remap,
     size_t worker_l1_size) {
-    ContextId context_id = MetalContext::create_instance(*this);
+    const bool env_owns_context = impl_->has_registered_context();
+    ContextId context_id =
+        env_owns_context ? ContextId{impl_->ensure_context_registered(*this)} : MetalContext::create_instance(*this);
     auto mesh_device = distributed::MeshDeviceImpl::create_unit_mesh(
         context_id,
         device_id,
@@ -658,7 +697,9 @@ std::shared_ptr<distributed::MeshDevice> MetalEnv::create_unit_mesh_device(
         dispatch_core_config,
         l1_bank_remap,
         worker_l1_size);
-    mesh_device->impl().set_destroy_metal_context_instance_on_close(true);
+    if (!env_owns_context) {
+        mesh_device->impl().set_destroy_metal_context_instance_on_close(true);
+    }
     return mesh_device;
 }
 
@@ -670,7 +711,9 @@ std::map<int, std::shared_ptr<distributed::MeshDevice>> MetalEnv::create_unit_me
     const DispatchCoreConfig& dispatch_core_config,
     ttsl::Span<const std::uint32_t> l1_bank_remap,
     size_t worker_l1_size) {
-    ContextId context_id = MetalContext::create_instance(*this);
+    const bool env_owns_context = impl_->has_registered_context();
+    ContextId context_id =
+        env_owns_context ? ContextId{impl_->ensure_context_registered(*this)} : MetalContext::create_instance(*this);
     auto result = distributed::MeshDeviceImpl::create_unit_meshes(
         context_id,
         device_ids,
@@ -680,7 +723,7 @@ std::map<int, std::shared_ptr<distributed::MeshDevice>> MetalEnv::create_unit_me
         dispatch_core_config,
         l1_bank_remap,
         worker_l1_size);
-    if (!result.empty()) {
+    if (!env_owns_context && !result.empty()) {
         const auto& parent = result.begin()->second->get_parent_mesh();
         if (parent) {
             parent->impl().set_destroy_metal_context_instance_on_close(true);

--- a/tt_metal/impl/context/metal_env_impl.hpp
+++ b/tt_metal/impl/context/metal_env_impl.hpp
@@ -90,6 +90,22 @@ public:
     // Teardown fabric-layer objects (control plane, system mesh, distributed context).
     void teardown_fabric_objects();
 
+    // Register this env as a MetalContext (silicon: default slot; mock: non-default slot) so that legacy
+    // fabric code reached during control-plane / system-mesh / mesh-device construction (which still calls
+    // the bare MetalContext::instance()) resolves to this env instead of implicitly creating a default
+    // silicon context that opens a second real PCIe device and self-deadlocks on its CHIP_IN_USE lock
+    // (GitHub #50041, #50043). Idempotent; returns the env-owned context id (created on first call). The
+    // context lifetime is tied to the env (uplift of mesh-device-level cleanup, GitHub #21500).
+    int ensure_context_registered(MetalEnv& env);
+
+    // True once ensure_context_registered has created the env-owned context.
+    bool has_registered_context() const { return registered_context_id_.has_value(); }
+
+    // Destroy the env-owned context created by ensure_context_registered, if any. Must be called from
+    // MetalEnv::~MetalEnv() while MetalEnv::impl_ is still valid (MetalContext::teardown() reaches back
+    // through *env_ -> impl_), not from ~MetalEnvImpl where impl_ has already been nulled by reset().
+    void teardown_registered_context();
+
     // Returns true if set_fabric_config changed state requiring a reinit.
     bool consume_force_reinit();
 
@@ -105,6 +121,11 @@ private:
     std::unique_ptr<Hal> hal_;
 
     std::atomic<int> use_count_{0};
+
+    // MetalContext id registered lazily for this env (see ensure_context_registered). Stored as a raw int
+    // to avoid pulling metal_context.hpp into this header; wrapped back into a ContextId at the
+    // metal_context.cpp boundary. Torn down by teardown_registered_context() from MetalEnv::~MetalEnv().
+    std::optional<int> registered_context_id_ = std::nullopt;
 
     // --- Fabric config state ---
     tt_fabric::FabricConfig fabric_config_ = tt_fabric::FabricConfig::DISABLED;

